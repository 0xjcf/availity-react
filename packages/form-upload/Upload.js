import React, { useRef, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import UploadCore from '@availity/upload-core';
import { avFilesDeliveryApi } from '@availity/api-axios';
import { Input, InputGroup } from 'reactstrap';
import { FormGroup, Feedback } from '@availity/form';
import Dropzone from 'react-dropzone';
import Icon from '@availity/icon';
import { useField, useFormikContext } from 'formik';
import classNames from 'classnames';
import uuid from 'uuid/v4';

import FilePickerBtn from './FilePickerBtn';
import FileList from './FileList';
import './styles.scss';

const Upload = ({
  allowedFileNameCharacters,
  allowedFileTypes,
  btnText,
  bucketId,
  children,
  className,
  clientId,
  customerId,
  deliverFileOnSubmit = false,
  deliveryChannel,
  disabled = false,
  feedbackClass,
  fileDeliveryMetadata,
  getDropRejectionMessage,
  max,
  maxSize,
  multiple = true,
  name,
  onFileRemove,
  onFileUpload,
  showFileDrop = false,
}) => {
  const input = useRef(null);
  const [field, metadata] = useField(name);
  const {
    errors,
    isSubmitting,
    isValidating,
    setFieldError,
    setFieldValue,
  } = useFormikContext();
  const classes = classNames(
    className,
    metadata.touched ? 'is-touched' : 'is-untouched',
    metadata.touched && metadata.error && 'is-invalid'
  );

  const fieldValue = Array.isArray(field.value) ? field.value : [];

  const callFileDelivery = useCallback(
    async upload => {
      if (!Array.isArray(upload)) upload = [upload];
      const uploadResults = [];
      try {
        for (const u of upload) {
          const data = {
            deliveries: [
              {
                deliveryChannel,
                fileURI: u.references[0],
                metadata:
                  typeof fileDeliveryMetadata === 'function'
                    ? fileDeliveryMetadata(u)
                    : fileDeliveryMetadata,
              },
            ],
          };

          uploadResults.push(
            avFilesDeliveryApi.uploadFilesDelivery(data, {
              clientId,
              customerId,
            })
          );
        }

        await Promise.all(uploadResults);
      } catch (error) {
        setFieldError(name, 'An error occurred while uploading files.');
      }
    },
    [
      clientId,
      customerId,
      deliveryChannel,
      fileDeliveryMetadata,
      name,
      setFieldError,
    ]
  );

  useEffect(() => {
    // eslint-disable-next-line unicorn/consistent-function-scoping
    async function checkValidFormAndCallFileDelivery() {
      if (Object.keys(errors).length === 0) {
        await callFileDelivery(fieldValue);
      }
    }
    if (
      !onFileUpload &&
      isSubmitting === true &&
      isValidating === false &&
      deliverFileOnSubmit &&
      deliveryChannel &&
      fileDeliveryMetadata
    ) {
      checkValidFormAndCallFileDelivery();
    }
  }, [
    callFileDelivery,
    deliverFileOnSubmit,
    deliveryChannel,
    errors,
    fieldValue,
    fileDeliveryMetadata,
    isSubmitting,
    isValidating,
    onFileUpload,
  ]);

  const removeFile = fileId => {
    const newFiles = fieldValue.filter(file => file.id !== fileId);
    if (newFiles.length !== fieldValue.length) {
      setFieldValue(name, newFiles, true);
      if (onFileRemove) onFileRemove(newFiles, fileId);
    }
  };

  const setFiles = files => {
    let selectedFiles = [];
    for (let i = 0; i < files.length; i++) {
      selectedFiles[i] = files[i];
    }

    if (max && selectedFiles.length + fieldValue.length > max) {
      selectedFiles = selectedFiles.slice(
        0,
        Math.max(0, max - fieldValue.length)
      );
    }

    const newFiles = fieldValue.concat(
      selectedFiles.map(file => {
        const upload = new UploadCore(file, {
          bucketId,
          customerId,
          clientId,
          fileTypes: allowedFileTypes,
          maxSize,
          allowedFileNameCharacters,
        });
        upload.id = `${upload.id}-${uuid()}`;
        if (file.dropRejectionMessage) {
          upload.errorMessage = file.dropRejectionMessage;
        } else {
          upload.start();
        }

        if (onFileUpload) {
          onFileUpload(upload);
        } else if (
          !deliverFileOnSubmit &&
          deliveryChannel &&
          fileDeliveryMetadata
        ) {
          upload.onSuccess.push(() => {
            callFileDelivery(upload);
          });
        }

        return upload;
      })
    );

    setFieldValue(name, newFiles, true);
  };

  const handleFileInputChange = event => {
    setFiles(event.target.files);
  };

  const onDrop = (acceptedFiles, fileRejections) => {
    const rejectedFilesToDrop = fileRejections.map(({ file, errors }) => {
      const dropRejectionMessage = getDropRejectionMessage
        ? getDropRejectionMessage(errors, file)
        : errors.map(error => error.message).join(', ');

      file.dropRejectionMessage = dropRejectionMessage;
      return file;
    });
    setFiles([...acceptedFiles, ...rejectedFilesToDrop]);
  };

  let fileAddArea;
  const text = btnText || (
    <>
      <Icon name="plus-circle" title="Add File Icon" />
      {fieldValue.length === 0 ? 'Add File' : 'Add Another File Attachment'}
    </>
  );

  if (!max || fieldValue.length < max) {
    if (showFileDrop) {
      fileAddArea = (
        <FormGroup for={name}>
          <Input name={name} style={{ display: 'none' }} />
          <InputGroup disabled={disabled} className={classes}>
            <Dropzone
              onDrop={onDrop}
              multiple={multiple}
              maxSize={maxSize}
              accept={allowedFileTypes}
            >
              {({ getRootProps, getInputProps, isDragActive }) => (
                <section>
                  <div
                    {...getRootProps({
                      className: isDragActive
                        ? 'file-drop-active'
                        : 'file-drop',
                    })}
                  >
                    <input data-testid="file-picker" {...getInputProps()} />
                    <p>
                      <strong>Drag and Drop</strong>
                    </p>
                    {text}
                  </div>
                </section>
              )}
            </Dropzone>
          </InputGroup>
          <Feedback
            className={classNames('d-block', feedbackClass)}
            name={name}
          />
        </FormGroup>
      );
    } else {
      fileAddArea = (
        <FilePickerBtn
          data-testid="file-picker"
          onChange={handleFileInputChange}
          color={fieldValue.length === 0 ? 'light' : 'link'}
          multiple={multiple}
          allowedFileTypes={allowedFileTypes}
          maxSize={maxSize}
          name={name}
          disabled={disabled}
        >
          {text}
        </FilePickerBtn>
      );
    }
  }

  return (
    <>
      <FileList files={fieldValue} onRemoveFile={removeFile}>
        {children}
      </FileList>
      {fileAddArea}
    </>
  );
};

Upload.propTypes = {
  allowedFileNameCharacters: PropTypes.string,
  allowedFileTypes: PropTypes.arrayOf(PropTypes.string),
  btnText: PropTypes.node,
  bucketId: PropTypes.string.isRequired,
  children: PropTypes.func,
  className: PropTypes.string,
  clientId: PropTypes.string.isRequired,
  customerId: PropTypes.string.isRequired,
  deliverFileOnSubmit: PropTypes.bool,
  deliveryChannel: PropTypes.string,
  disabled: PropTypes.bool,
  feedbackClass: PropTypes.string,
  fileDeliveryMetadata: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
  getDropRejectionMessage: PropTypes.func,
  max: PropTypes.number,
  maxSize: PropTypes.number,
  multiple: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onFileRemove: PropTypes.func,
  onFileUpload: PropTypes.func,
  showFileDrop: PropTypes.bool,
};

export default Upload;
