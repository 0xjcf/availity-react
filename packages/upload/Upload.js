import React, { Component, createRef, Fragment } from 'react';
import PropTypes from 'prop-types';
import UploadCore from '@availity/upload-core';
import FilePickerBtn from './FilePickerBtn';
import FileList from './FileList';

const validationAttrs = ['min', 'max', 'required'];

class Upload extends Component {
  static propTypes = {
    btnText: PropTypes.node,
    bucketId: PropTypes.string.isRequired,
    customerId: PropTypes.string.isRequired,
    clientId: PropTypes.string.isRequired,
    allowedFileTypes: PropTypes.arrayOf(PropTypes.string),
    onFileUpload: PropTypes.func,
    onFileRemove: PropTypes.func,
    maxSize: PropTypes.number,
    max: PropTypes.number,
    multiple: PropTypes.bool,
    children: PropTypes.func,
    name: PropTypes.string,
  };

  static defaultProps = {
    multiple: true,
  };

  input = createRef();

  files = [];

  state = {
    files: [],
  };

  removeFile = fileId => {
    this.setState(({ files }) => {
      const newFiles = files.filter(file => file.id !== fileId);
      if (newFiles.length !== files.length) {
        this.files = newFiles;
        if (this.props.onFileRemove) this.props.onFileRemove(this.files);
        return {
          files: newFiles,
        };
      }
      return null;
    });
  };

  handleFileInputChange = event => {
    const { files: selectedFilesList } = event.target;
    let selectedFiles = [];
    for (let i = 0; i < selectedFilesList.length; i++) {
      selectedFiles[i] = selectedFilesList[i];
    }
    if (
      this.props.max &&
      selectedFiles.length + this.state.files.length > this.props.max
    ) {
      selectedFiles = selectedFiles.slice(
        0,
        Math.max(0, this.props.max - this.state.files.length)
      );
    }
    this.files = this.files.concat(
      selectedFiles.map(file => {
        const upload = new UploadCore(file, {
          bucketId: this.props.bucketId,
          customerId: this.props.customerId,
          clientId: this.props.clientId,
          fileTypes: this.props.allowedFileTypes,
          maxSize: this.props.maxSize,
        });
        upload.start();
        if (this.props.onFileUpload) this.props.onFileUpload(upload);
        return upload;
      })
    );
    this.setState({ files: this.files });
  };

  reset = () => {
    this.files = [];
    this.setState({ files: [] });
  };

  componentDidMount() {
    if (this.context.FormCtrl && this.props.name) {
      this.updateValidations();
    }
  }

  updateValidations(props = this.props) {
    this.validations = Object.assign({}, props.validate);

    Object.keys(props)
      .filter(val => validationAttrs.indexOf(val) > -1)
      .forEach(attr => {
        if (props[attr]) {
          this.validations[attr] = this.validations[attr] || {
            value: props[attr],
          };
        } else {
          delete this.validations[attr];
        }
      });

    this.context.FormCtrl.register(this);
    this.validate();
  }

  validate() {
    if (this.context.FormCtrl && this.props.name) {
      this.context.FormCtrl.validate(this.props.name);
    }
  }

  componentWillUnmount() {
    if (this.context.FormCtrl && this.props.name)
      this.context.FormCtrl.unregister(this);
  }

  getValue() {
    if (!this.files) return [];
    return this.files;
  }

  render() {
    const {
      btnText,
      max,
      multiple,
      allowedFileTypes,
      maxSize,
      children,
    } = this.props;
    const { files } = this.state;
    return (
      <Fragment>
        <FileList files={files} onRemoveFile={this.removeFile}>
          {children}
        </FileList>
        {(!max || files.length < max) && (
          <FilePickerBtn
            onChange={this.handleFileInputChange}
            color={files.length === 0 ? 'light' : 'link'}
            multiple={multiple}
            allowedFileTypes={allowedFileTypes}
            maxSize={maxSize}
          >
            {btnText || (
              <Fragment>
                <i className="icon icon-plus-circle" title="Add File Icon" />
                {files.length === 0
                  ? 'Add File'
                  : 'Add Another File Attachment'}
              </Fragment>
            )}
          </FilePickerBtn>
        )}
      </Fragment>
    );
  }
}

export default Upload;
