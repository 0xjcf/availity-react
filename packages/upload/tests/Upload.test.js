import React from 'react';
import { screen, render, fireEvent, waitFor } from '@testing-library/react';

import Upload from '..';

const getDropRejectionMessage = (errors) => {
  let msg = '';
  for (const error of errors) {
    msg += error.code === 'file-too-large' ? 'my custom error message' : 'this file is no good';
  }
  return msg;
};

describe('Upload', () => {
  test('should render', () => {
    const { container } = render(<Upload clientId="a" bucketId="b" customerId="c" />);

    expect(container).toMatchSnapshot();
  });

  test('adding a file', () => {
    render(<Upload clientId="a" bucketId="b" customerId="c" showFileDrop />);

    const file = Buffer.from('hello world');
    file.name = 'fileName.png';
    const fileEvent = { target: { files: [file] } };

    const inputNode = screen.getByTestId('file-picker');

    fireEvent.change(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
  });

  test('removing a file', () => {
    render(<Upload clientId="a" bucketId="b" customerId="c" />);

    // Create a new file
    const file = Buffer.from('hello world');
    file.name = 'fileName.png';

    const fileEvent = { target: { files: [file] } };

    const inputNode = screen.getByTestId('file-picker');

    // Simulate the upload to the Components
    fireEvent.change(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);

    const filerow = screen.getByTestId('remove-file-btn');

    fireEvent.click(filerow);

    expect(screen.queryByTestId('remove-file-btn')).toBeNull();
  });

  test('calls onFilePreUpload callback', () => {
    const mockFunc = jest.fn();
    render(<Upload clientId="a" bucketId="b" customerId="c" onFilePreUpload={[mockFunc]} />);
    const inputNode = screen.getByTestId('file-picker');

    const file = Buffer.from('hello world');
    file.name = 'fileName.png';
    const fileEvent = { target: { files: [file] } };

    fireEvent.change(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);

    const filerow = screen.getByTestId('remove-file-btn');

    fireEvent.click(filerow);

    expect(mockFunc).toHaveBeenCalled();
  });

  test('calls onFileRemove callback', () => {
    const mockFunc = jest.fn();

    render(<Upload clientId="a" bucketId="b" customerId="c" onFileRemove={mockFunc} />);

    const inputNode = screen.getByTestId('file-picker');

    const file = Buffer.from('hello world');
    file.name = 'fileName.png';
    const fileEvent = { target: { files: [file] } };

    fireEvent.change(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);

    const filerow = screen.getByTestId('remove-file-btn');

    fireEvent.click(filerow);

    expect(mockFunc).toHaveBeenCalled();
  });

  test('adds file via dropzone', () => {
    render(<Upload clientId="a" bucketId="b" customerId="c" showFileDrop />);
    const file = Buffer.from('hello world');
    file.name = 'fileName.png';

    const inputNode = screen.getByTestId('file-picker');
    const fileEvent = { target: { files: [file] } };

    fireEvent.drop(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
  });

  test('uses default drop rejection message', async () => {
    render(<Upload clientId="a" bucketId="b" customerId="c" showFileDrop maxSize={10} />);
    const file = Buffer.from('hello world');
    file.name = 'fileName.png';
    file.size = 11;

    const inputNode = screen.getByTestId('file-picker');
    const fileEvent = { target: { files: [file] } };

    fireEvent.drop(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
    await waitFor(() => {
      expect(screen.getByText('File is larger than 10 bytes')).toBeDefined();
    });
  });

  test('a discontinue result from a function in the property onFilePreUpload should prevent onFileUpload from being called', async () => {
    const myfunc = jest.fn();
    render(
      <Upload
        clientId="a"
        bucketId="b"
        customerId="c"
        showFileDrop
        maxSize={10}
        onFileUpload={myfunc}
        onFilePreUpload={[() => false]}
      />
    );
    const file = Buffer.from('hello world');
    file.name = 'fileName.png';
    file.size = 11;
    const inputNode = screen.getByTestId('file-picker');
    const fileEvent = { target: { files: [file] } };

    fireEvent.drop(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
    expect(myfunc).toHaveBeenCalledTimes(0);
  });

  test('a discontinue result from a function in the property onFilePreUpload does not stop an error message from generating', async () => {
    render(
      <Upload
        clientId="a"
        bucketId="b"
        customerId="c"
        showFileDrop
        maxSize={10}
        getDropRejectionMessage={getDropRejectionMessage}
        onFilePreUpload={[() => false]}
      />
    );
    const file = Buffer.from('hello world');
    file.name = 'fileName.png';
    file.size = 11;
    const inputNode = screen.getByTestId('file-picker');
    const fileEvent = { target: { files: [file] } };

    fireEvent.drop(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
    await waitFor(() => {
      expect(screen.getByText('my custom error message')).toBeDefined();
    });
  });

  test('uses custom drop rejection message', async () => {
    render(
      <Upload
        clientId="a"
        bucketId="b"
        customerId="c"
        showFileDrop
        maxSize={10}
        getDropRejectionMessage={getDropRejectionMessage}
      />
    );
    const file = Buffer.from('hello world');
    file.name = 'fileName.png';
    file.size = 11;

    const inputNode = screen.getByTestId('file-picker');
    const fileEvent = { target: { files: [file] } };

    fireEvent.drop(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
    await waitFor(() => {
      expect(screen.getByText('my custom error message')).toBeDefined();
    });
  });

  test('uses cloud url when isCloud is true', async () => {
    const mockFn = jest.fn();

    render(
      <Upload
        clientId="a"
        bucketId="b"
        customerId="c"
        isCloud
        onFilePreUpload={[
          (file) => {
            mockFn(file.options.endpoint);
          },
        ]}
      />
    );

    const file = Buffer.from('hello world');
    file.name = 'fileName.png';
    const fileEvent = { target: { files: [file] } };

    const inputNode = screen.getByTestId('file-picker');

    fireEvent.change(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);

    await waitFor(() => {
      expect(mockFn).toHaveBeenCalledWith('http://localhost/cloud/vault/upload/v1/resumable');
    });
  });
});
