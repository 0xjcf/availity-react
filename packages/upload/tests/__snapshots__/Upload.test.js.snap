// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload should render 1`] = `
Array [
  <div
    className="d-none"
  >
    <input
      className="form-control-file"
      id="filePicker-1"
      multiple={true}
      onChange={[Function]}
      type="file"
      value=""
    />
  </div>,
  <button
    className="btn btn-light"
    onClick={[Function]}
    type="button"
  >
    <i
      className="icon icon-plus-circle"
      title="Add File Icon"
    />
    Add File
  </button>,
]
`;
